src/standalone/diagnostics/reservedFileNameDiagnostics.node.ts

```diff
@@ -0,0 +1,208 @@
+import { swallowExceptions } from '../../platform/common/utils/decorators';
+
+export const enabledSettingName = 'diagnostics.reservedPythonNames.enabled';
+
+@injectable()
+export class ReservedFileNamesDiagnosticProvider
+    implements IExtensionSingleActivationService, CodeActionProvider, FileDecorationProvider
+{
+    private readonly disposables: IDisposable[] = [];
+    private readonly diagnosticCollection: DiagnosticCollection;
+    private readonly _onDidChangeFileDecorations = new EventEmitter<Uri | Uri[] | undefined>();
+    onDidChangeFileDecorations = this._onDidChangeFileDecorations.event;
+    private enabled: boolean;
+    constructor(
+        @inject(IReservedPythonNamedProvider) private readonly reservedNameProvider: IReservedPythonNamedProvider,
+        @inject(IFileSystem) private readonly fileSystem: IFileSystem,
+        @inject(IWorkspaceService) private workspace: IWorkspaceService
+    ) {
+        this.diagnosticCollection = languages.createDiagnosticCollection(
+            DataScience.reservedPythonFileNamesDiagnosticCollectionName()
+        );
+    }
+    public dispose() {
+        disposeAllDisposables(this.disposables);
+        this._onDidChangeFileDecorations.dispose();
+        this.diagnosticCollection.dispose();
+    }
+    public async activate(): Promise<void> {
+        this.disposables.push(languages.registerCodeActionsProvider(PYTHON_LANGUAGE, this));
+        this.disposables.push(window.registerFileDecorationProvider(this));
+        this.enabled = this.workspace.getConfiguration('jupyter').get<boolean>(enabledSettingName, true);
+        this.workspace.onDidChangeConfiguration(
+            (e) => {
+                if (e.affectsConfiguration(`jupyter.${enabledSettingName}`)) {
+                    this.enabled = this.workspace.getConfiguration('jupyter').get<boolean>(enabledSettingName, true);
+                    if (this.enabled) {
+                        window.visibleTextEditors.forEach((editor) => this.provideDiagnosticsForEditor(editor));
+                    } else {
+                        this.diagnosticCollection.clear();
+                    }
+                    this._onDidChangeFileDecorations.fire(undefined);
+                }
+            },
+            this,
+            this.disposables
+        );
+
+        window.onDidChangeActiveTextEditor(this.provideDiagnosticsForEditor, this, this.disposables);
+        workspace.onDidCloseTextDocument((e) => this.diagnosticCollection.delete(e.uri), this, this.disposables);
+        window.visibleTextEditors.forEach((editor) => this.provideDiagnosticsForEditor(editor));
+        this.disposables.push(
+            commands.registerCommand(
+                'jupyter.ignoreReservedFileNamesDiagnostic',
+                async (uri: Uri) => {
+                    await this.reservedNameProvider.addToIgnoreList(uri);
+                    this._onDidChangeFileDecorations.fire(uri);
```

Comment:
Ordering might be off here. I would think you'd fire the deco change after deleting.

Reply:
No its correct, because we basically tell VS Code the decorations have changed for the file and when vscode calls back into our code to get the decorations we should have ignored the file (meaning that should have happened first)